### DISCUSSION TITLE
Suppress Casadi logging entirely?

### DISCUSSION THREAD

**Discussion by bardsleypt** [pybamm-team/PyBaMM#2102] [Q&A] [answered]
I am using a (relatively) simple model+experiment+params+solver combination inside of an objective function, all with the goal of optimizing a particular parameter set. In this case, the optimization evaluates the objective function numerous times, and many evaluation positions are infeasible/bad leading to Casadi warnings/errors. I've wrapped this objective evaluation in a try/except that suits my purposes, but the Casadi outputs are overly verbose (see below):

<img width="868" alt="Screen Shot 2022-06-14 at 12 24 14 PM" src="https://user-images.githubusercontent.com/54084289/173661926-9668d539-9496-4a43-9032-2def6cecbe9b.png">
 
I understand these warnings are wanted/needed in a lot of cases, but not in mine - so I'm curious if there is an option I can flag somewhere to completely suppress this logging output? I've searched the documentation, the most likely solution being setting the correct 'extra_options_call' option when setting up the solver (e.g., disable_internal_warnigns), but this did not work. I also looked through other discussions but couldn't find a similar question (apologies if I missed something).

If this option isn't available, could I suggest it as a feature in the future?

Thanks in advance.

**Comment by valentinsulzer** [comment, marked as answer]
Yeah, we've never figured out how to do this. It seems there should be some brute force way to divert the terminal output. Would be awesome if you can figure it out.

**Reply by bardsleypt** [reply]
Okay, thanks for the info. If I find a way, I'll be sure to contribute.

**Comment by martinjrobins** [comment]
depending on how Casadi does its logging, you might want to try `redirect_stdout` and/or `redirect_stderr` in contextlib: https://docs.python.org/3/library/contextlib.html

**Reply by bardsleypt** [reply]
great, thanks for the suggestion. I'll look this direction for a workaround.

**Comment by leesh1336** [comment]
I'm currently studying PyBaMM, and I have a exactly same question to yours.
I'm working on a code to optimize electrochemical parameters for a battery simulation model. 
Could you please provide some guidance on how you managed this problem? 
I understand that revisiting code from two years ago may not be easy, but I've been struggling with this issue for almost two weeks now. Thank you!

**Reply by bardsleypt** [reply]
Hi @leesh1336.

I never found a good way of turning off Casadi logging altogether, but I at least found a way of suppressing it using a context manager as suggested above:

```
import os
import sys
from contextlib import redirect_stdout, redirect_stderr, contextmanager

@contextmanager
def silence_output():
    with open(os.devnull, 'w') as devnull:
        with redirect_stdout(devnull) as devnull, redirect_stderr(sys.stdout):
            yield

# call the noisy simulation
with silence_output():
    sol = pybamm.Simulation(**sim_kwargs).solve(**sol_kwargs)
```

This quieted down the logging for my purposes at the time. This comes with the standard disclaimer: I haven't used this code in over 2 years, so there may be improvements/fixes that I'm not aware of. But I hope this helps.

Best,
Patrick

**Reply by leesh1336** [reply]
Dear, @bardsleypt 

Thank you for your kind reply. 
Recalling the code from two years ago, as you mentioned, is not easy. 

Your code has been very helpful for me.
You've relieved me of my work a lot

Thank you very much,
Best,
Seunghyun

**Reply by bardsleypt** [reply]
you're welcome, happy to hear that it helped!
